### PR TITLE
Fixes #4692 and #4594

### LINKED ISSUE #4692: jax.scipy.minimize optional function args not being passed properly
In `jax.scipy.optimize._minimize`,
```python
fun_with_args = partial(fun, *args)
```
incorrectly puts the `args` before the parameter when `args` are provided.
E.g. 

```python
from functools import partial
func = lambda a,b: print(a,b)

args = ('b',)
partial(func, *args)('a')
# b a instead of a b as intended
```

### PR DESCRIPTION
**Pull request by Joshuaalbert**
Minor code changes to fix #4692 and #4594

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi - thanks for the contribution! I think the fix for #4692 is fine, but I'm not sure about the change related to #4594. How did you come up with `1E-5` here, as opposed to, say, `1E-6` or `1E-4`? Is there a test we could add that would catch regressions along the lines of the issue reported in #4594?

All told, I think it might be beneficial to contribute these changes separately, since they require distinct discussions.

**Comment by Joshuaalbert** [comment]
Hi @jakevdp, when I wrote the initial line search method I used a threshold for zooming such that if too-small step sizes are tried then it stops. This typically means you're at a saddle point. I used the threshold of 1e-5 because that worked well in the tests I used, and indeed all the comparitive tests with scipy in the unit tests pass with 1e-5. However 1e-5 is not small enough when the jacobian large near the initial variable of a minimisation. That was the issue that arose in #4594. I've now chosen a value of 1e-10 to allow less well behaved objectives near the initial point. Note, that after a few iterations of BFGS or any other approximate variable metric method the step size will settle around unity so you don't need to worry about small lengthscales, except when you hit a saddle point. The unit tests are still passing because this change doesn't impact cases with well-behaved jacobians near the initial point. I'm happy to revert 1e-10 to 1e-5.

**Comment by jakevdp** [comment]
Can you add a test case for the issue in #4594; i.e. one that passes with `1E-10` but fails with `1E-5`?

**Comment by jakevdp** [comment]
Looks great, thanks!

One last thing: could you please squash these changes into a single commit? In the meantime, I'll trigger GPU and TPU versions of the new test.

**Comment by Joshuaalbert** [comment]
squashed

**Review by jakevdp — APPROVED** [review]

File: tests/scipy_optimize_test.py [hunk 1]
```diff
@@ -86,6 +86,14 @@ def min_op(x0):
     scipy_res = scipy.optimize.minimize(func(np), x0, method='BFGS').x
     self.assertAllClose(scipy_res, jax_res, atol=2e-5, check_dtypes=False)
 
+  def test_fixes4594(self):
+    n = 2
+    A = jnp.eye(n) * 1e4
+    def f(x):
+      return jnp.mean((A @ x) ** 2)
+    results = jax.scipy.optimize.minimize(f, jnp.ones(n), method='BFGS')
+    assert jnp.isclose(results.x, jnp.zeros(n)).all()
```
Comment (jakevdp) [inline tests/scipy_optimize_test.py:95]: This test fails on TPU; we'll need to change the precision

Also, since this is in the test framework, we should be using `self.assertAllClose`. If it had been using that, I'd have been able to tell you what precision we need 😀

Comment (Joshuaalbert) [reply tests/scipy_optimize_test.py]: Oops, I changed it now to try with 1e-8.

File: tests/scipy_optimize_test.py [hunk 2]
```diff
@@ -86,6 +86,14 @@ def min_op(x0):
     scipy_res = scipy.optimize.minimize(func(np), x0, method='BFGS').x
     self.assertAllClose(scipy_res, jax_res, atol=2e-5, check_dtypes=False)
 
+  def test_fixes4594(self):
+    n = 2
+    A = jnp.eye(n) * 1e4
+    def f(x):
+      return jnp.mean((A @ x) ** 2)
+    results = jax.scipy.optimize.minimize(f, jnp.ones(n), method='BFGS')
+    self.assertAllClose(results.x, jnp.zeros(n), atol=1e-8)
```
Comment (jakevdp) [inline tests/scipy_optimize_test.py:95]: Still fails on TPU. Let's switch this to `atol=1e-6, rtol=1e-6`, then squash into a single commit, and I think we'll be good.

Comment (Joshuaalbert) [reply tests/scipy_optimize_test.py]: done

**Review by jakevdp — APPROVED** [review]
